#46252 First pass of multi context introspection module.

manneohrstrom opened the pull request:
Sphinx documentation bake to show the interface:

![image](https://user-images.githubusercontent.com/337710/36424835-a8f040c6-163c-11e8-9cac-4d60901e1df7.png)

This adds a new module to shotgun utils to help handle cross configuration introspection and launch. For intro, see the docs above. There is still stuff to do:

- Unit tests will be handled as a separate ticket once the interfaces have settled
- QT based external script running not in place yet.

manneohrstrom commented on python/external_config/process_execution.py:
```diff
@@ -0,0 +1,260 @@
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
```
note: this was carried over from the server fw (which carried it over from the desktop engine). We may not need this in the future at all, but it's there for the time being.

manneohrstrom commented:
Note: Since this is still WIP with some more changes anticipated and tests due, i have created a `feature/external_config` branch which we will merge this into once it's been through CR. We can then release early RC type cuts from that branch.

sgsteph commented on python/external_config/scripts/external_runner.py:
```diff
@@ -0,0 +1,221 @@
+        logger.exception("Could not bootstrap configuration")
+
+        # print to our logger so it's picked up by the main process
+        print traceback.format_exc()
```
Seems strange to have a `print` here, why not having a handler to the logger which logs on stdout?

manneohrstrom replied:
i'll take a look. this is old code. Not sure a logger is needed. We only want to pass across error messages.

sgsteph commented on python/external_config/scripts/external_runner.py:
```diff
@@ -0,0 +1,221 @@
+
+    except EngineStartupFailure:
+        sys.exit(2)
+    except Exception, e:
```
`as e`?

sgsteph commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+        if sg_data is None:
+            return None
+        else:
+            return hash(str(sg_data))
```
Possibly a totally stupid question, sorry, but why do we need to compute such a hash based on the data? If the goal is to know something has changed in the data, we have signals which could be used to set a hash value based on the `data_changed` bool (that you fixed), possibly based on a timestamp?

manneohrstrom replied:
agree this is hacky. i'll fix it as part of the hash cleanup

manneohrstrom commented on framework.py:
```diff
@@ -24,6 +24,7 @@ class ShotgunUtilsFramework(sgtk.platform.Framework):
         "sg",
         "sg_nav",
         "thumbs",
+        "multi_context"
```
yeah!!! soooooooo awesome!

manneohrstrom commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+        configuration state. If a change is detected, indicating that
+        configurations should be recomputed, a ``state_changed`` signal is emitted.
+        """
+        self._pipeline_config_model.load()
```
Great point! fixed!

manneohrstrom commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+        pc_hash = self._pipeline_config_model.get_hash()
+        if sw_hash is None or pc_hash is None:
+            return None
+        return str(sw_hash) + str(pc_hash)
```
fixed. this is also part of a bigger overhaul/centralization of the caching logic which we should do.

manneohrstrom commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+        :returns: Hash string or ``None`` if not yet defined.
+        """
+        sw_hash = self._software_model.get_hash()
+        pc_hash = self._pipeline_config_model.get_hash()
```
fixed.

manneohrstrom commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+
+class SoftwareModel(ShotgunModel):
+    """
+    All software entities
```
good point! fixed!

manneohrstrom commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+        """
+        Currently loaded Shotgun data.
+
+        :returns: The sg data dictionary for the associated item,
```
fixed!

manneohrstrom commented on python/external_config/configuration_state.py:
```diff
@@ -0,0 +1,219 @@
+        return data
+
+
+class PipelineConfigModel(ShotgunModel):
```
Yes, that's probably cleaner! I'll rejig.

manneohrstrom commented on python/external_config/file_cache.py:
```diff
@@ -0,0 +1,155 @@
+        try:
+            with open(cache_path, "rb") as fh:
+                content = pickle.load(fh)
+        except Exception, e:
```
yes, sorry bad habits

manneohrstrom replied:
changed everywhere.

manneohrstrom commented on python/external_config/file_cache.py:
```diff
@@ -0,0 +1,155 @@
+            pickle.dump(data, fh)
+
+        # and ensure the cache file has got open permissions
+        os.chmod(path, 0666)
```
yes, i *think* so, but not 100% sure. i'll check it as part of the cache refactors.

manneohrstrom commented on python/external_config/file_cache.py:
```diff
@@ -0,0 +1,155 @@
+    except Exception, e:
+        logger.debug("Could not write '%s'. Details: %s" % (path, e))
+
+    logger.debug("Completed save of %s. Size %s bytes" % (path, os.path.getsize(path)))
```
fixed!

manneohrstrom commented on python/external_config/file_cache.py:
```diff
@@ -0,0 +1,155 @@
+        os.chmod(path, 0666)
+
+    except Exception, e:
+        logger.debug("Could not write '%s'. Details: %s" % (path, e))
```
awesome! i'll get into the habit of doing that.

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+CONFIGURATION_GENERATION = 1
+
+
+class RemoteCommand(object):
```
all changed to external

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        This can be passed to :meth:`create` in order to construct a
+        :class:`RemoteCommand` instance.
+
+        :param dict entity: Shotgun entity dictionary with keys type and id.
```
switched this over to be entity type instead, and clarified.

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        )
+
+    @classmethod
+    def from_string(cls, data):
```
good point!

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        data = cPickle.loads(data)
+
+        if data.get("GENERATION") != CONFIGURATION_GENERATION:
+            raise RuntimeError("Format is incompatible.")
```
ok!

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        :rtype: :class:`RemoteCommand`
+        :raises: :class:`RuntimeError` if data is not valid
+        """
+        data = data.encode("utf-8")
```
good point, it should be in the desktop2 engine!

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+    def pipeline_configuration_name(self):
+        """
+        The name of the Shotgun pipeline configuration this command is associated with,
+        or ``None`` if no association exists.
```
yep, for the most basic zero config case when a user doesn't have any pipeline configurations associated.

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        .. note:: The process will be launched in an asynchronous way.
+            It is recommended that this command is executed in a worker thread.
+        """
+        # local imports becuase this is executed from runner scripts
```
fixed

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        """
+        Executes the remote command in a separate process.
+
+        .. note:: The process will be launched in an asynchronous way.
```
this needs a bit further work once i can run the publisher. I am not sure that process runner class is needed yet. most of this code was carried across from the desktop server. stay tuned!

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        ]
+        logger.debug("Command arguments: %s", args)
+
+        retcode, stdout, stderr = ProcessRunner.call_cmd(args)
```
we discussed this - but it may be tricky because each process is likely to require it's own sgtk environment (or it *might* be possible with core swapping, but certainly safer this way). I think it's about finding the right tradeoff between performance and complexity. Let me do the launch publisher/run QT in the external process tests and then i think we will know more.

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+            logger.error("Failed command stdout: %s", stdout)
+            logger.error("Failed command stderr: %s", stderr)
+            logger.error("Failed command retcode: %s", retcode)
+            raise Exception("%s\n\n%s" % (stdout, stderr))
```
yeah, old copy paste code. will fix.

manneohrstrom commented on python/external_config/remote_command.py:
```diff
@@ -0,0 +1,288 @@
+        retcode, stdout, stderr = ProcessRunner.call_cmd(args)
+
+        if retcode == 0:
+            logger.error("Command stdout: %s", stdout)
```
fixed!

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,300 @@
+# Copyright (c) 2018 Shotgun Software Inc.
```
normally yes! but i feel it's clearer to browse this way when there are deriving classes. You end up with this strucure, which is grouped together and clear:

![image](https://user-images.githubusercontent.com/337710/36535487-bdc81cf4-17c1-11e8-9721-e250b177f0f9.png)

Hope it makes sense!

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,300 @@
+        """
+        super(RemoteConfiguration, self).__init__(parent)
+
+        self._parent = parent
```
no idea why it was there. part of an earlier iteration. gone!

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        """
+        The associated pipeline configuration id or ``None`` if not defined.
+        """
+        return None
```
this is subclased and those classes may override the value.

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        """
+        The name of the associated pipeline configuration or ``None`` if not defined.
+        """
+        return None
```
this is subclased and those classes may override the value. i'll add comments.

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        configurations that have an associated :meth:`pipeline_configuration_id`,
+        this returns ``None``.
+        """
+        return None
```
yes! done!

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+
+    def request_commands(self, entity_type, entity_id, link_entity_type):
+        """
+        Request commands for the given object.
```
changed

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        commands parameter contains a list of :class:`RemoteCommand` instances.
+
+    """
+    TASK_GROUP = "tk-framework-shotgunutils.external_config.RemoteConfiguration"
```
done

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+
+    def _compute_config_hash(self, entity_type, entity_id, link_entity_type):
+        """
+        Generates a hash to uniquely identify the configuration.
```
fixed!

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+            where caching it per linked type can be beneficial.
+        """
+        # figure out if we have a suitable config for this on disk already
+        cache_hash = self._compute_config_hash(entity_type, entity_id, link_entity_type)
```
done

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+            except SubprocessCalledProcessError, e:
+                # caching failed!
+                logger.error("External process command caching failed: %s" % e.output)
+                raise Exception("Error retrieving actions.")
```
This will be reported back to the user so its meant to be more user friendly.

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+                subprocess_check_output(args)
+            except SubprocessCalledProcessError, e:
+                # caching failed!
+                logger.error("External process command caching failed: %s" % e.output)
```
we don't want that in this case - we want to include the output from the script which failed!

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        if cached_data:
+            self.commands_loaded.emit(
+                self,
+                [RemoteCommand.create(self, d, entity_id) for d in cached_data]
```
yep good point, this was an optimization to get the I/O to load in the thread. I'll reshuffle to make sure QObjects are born in the main thread.

manneohrstrom replied:
fixed

manneohrstrom commented on python/external_config/remote_config/config_fallback.py:
```diff
@@ -0,0 +1,88 @@
+            "link_type": link_entity_type,
+        }
+
+        return cache_key
```
yes you only need to say it once. It's changed already.

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        del self._task_ids[unique_id]
+
+        # log exception message to error log
+        logger.error(message)
```
it's being logged, let me see if i can make this clearer.

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        # log exception message to error log
+        logger.error(message)
+        # emit an empty list of commands
+        self.commands_loaded.emit(self, [])
```
`_task_completed` now emits stuff too so that everything gets emitted from the main thread

manneohrstrom commented on python/external_config/remote_config/config_fallback.py:
```diff
@@ -0,0 +1,88 @@
+        :returns: dictionary of values to use for hash computation
+        """
+        cache_key = {
+            file_cache.FOLDER_PREFIX_KEY: "base",
```
yes possibly. i'll look during cache refactor.

manneohrstrom commented on python/external_config/remote_config/config_base.py:
```diff
@@ -0,0 +1,285 @@
+        :param bg_task_manager: Background task manager to use for any asynchronous work.
+        :type bg_task_manager: :class:`~task_manager.BackgroundTaskManager`
+        :param str plugin_id: Associated bootstrap plugin id
+        :param str engine: Associated engine name
```
done everywhere.

manneohrstrom commented on python/external_config/remote_config/config_fallback.py:
```diff
@@ -0,0 +1,88 @@
+class FallbackRemoteConfiguration(RemoteConfiguration):
+    """
+    Class representing a config which does not have
+    an associated pipeline configuration id.
```
done

manneohrstrom commented on python/external_config/remote_config/config_fallback.py:
```diff
@@ -0,0 +1,88 @@
+        """
+        cache_key = {
+            file_cache.FOLDER_PREFIX_KEY: "base",
+            "engine": self.engine,
```
fixed.

manneohrstrom commented on python/external_config/remote_config/config_immutable.py:
```diff
@@ -0,0 +1,104 @@
+    @property
+    def pipeline_configuration_id(self):
+        """
+        The associated pipeline configuration id or ``None`` if not defined.
```
yep for fallback configs

manneohrstrom commented on python/external_config/remote_config/config_live.py:
```diff
@@ -0,0 +1,159 @@
+        )
+
+    @property
+    def pipeline_configuration_id(self):
```
i try to avoid deep class hierarchies. I think in this case we should keep it the way it is.

manneohrstrom commented on python/external_config/remote_config/config_live.py:
```diff
@@ -0,0 +1,159 @@
+                ...
+            }
+
+        :returns: A dictionary keyed by yml file path, set to the file's mtime
```
made changes and didn't update the docstring sorry. fixed now.

manneohrstrom commented on python/external_config/remote_config/config_live.py:
```diff
@@ -0,0 +1,159 @@
+        return cache_key
+
+    @sgtk.LogManager.log_timing
+    def _get_yml_file_data(self):
```
yep, this is old code i pulled across. now `_get_environment_hash`

manneohrstrom commented on python/external_config/remote_config/factory.py:
```diff
@@ -0,0 +1,184 @@
+    :param configuration_data: Dictionary entry on the form
+        returned by ToolkitManager.get_pipeline_configurations()
+    :returns: :class:`RemoteConfiguration`
+    :raises: :class:`RemoteConfigNotAccessibleError` if the configuration data could not
```
fixed

manneohrstrom commented on python/external_config/remote_config/factory.py:
```diff
@@ -0,0 +1,184 @@
+        )
+
+
+def create_default(parent, bg_task_manager, config_loader):
```
renamed.

manneohrstrom commented on python/external_config/remote_config/factory.py:
```diff
@@ -0,0 +1,184 @@
+            data["pipeline_config_name"],
+            data["config_uri"],
+        )
+    elif data["class_name"] == "LiveRemoteConfiguration":
```
in my experience that kind of code is more bug prone than the one that's currently there.

manneohrstrom commented on python/external_config/remote_config/factory.py:
```diff
@@ -0,0 +1,184 @@
+            data["interpreter"],
+            data["config_uri"],
+        )
+    else:
```
in my experience, i have an else for every if unless there is a special situation. This seems to minimize bugs over time.

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+        project id and a list of :class:`RemoteConfiguration` instances.
+
+    :signal configurations_changed(): Gets emitted whenever the class
+        has detected a change to the state of shotgun which could invalidate
```
done

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+    # signal to indicate that change to the configurations have been detected.
+    configurations_changed = QtCore.Signal()
+
+    TASK_GROUP = "tk-framework-shotgunutils.external_config.RemoteConfigurationLoader"
```
done

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+        .. note:: The interpreter needs to support the VFX Platform, e.g be
+            able to import ``PySide`` or ``Pyside2``.
+
+        :param str interpreter: Path to python interpreter to use.
```
COME ON!

manneohrstrom commented on python/external_config/scripts/external_runner.py:
```diff
@@ -0,0 +1,221 @@
+        )
+        logger.debug("Engine %s started using entity %s %s", engine, entity_type, entity_id)
+
+    except Exception, e:
```
done (you only need to point it out once in the CR if that makes it easier, and then i'll do a big search replace :) )

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+                self.configurations_loaded.emit(project_id, config_objects)
+                config_data_emitted = True
+
+        if not config_data_emitted:
```
tried it but i feel it's clearer the way it is.

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+        config_data = file_cache.load_cache(config_cache_key)
+        # attempt to load configurations
+        config_data_emitted = False
+        if config_data and not force:
```
changed to `force_reload`

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+                remote_config.create_default(
+                    self,
+                    self._bg_task_manager,
+                    self
```
now passing parms by name for clarity

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+        data = {
+            "project_id": project_id,
+            "plugin_id": self._plugin_id,
+            "global_state_hash": hash,
```
renamed everywhere to `state_hash`

manneohrstrom commented on python/external_config/remote_config_loader.py:
```diff
@@ -0,0 +1,279 @@
+        del self._task_ids[unique_id]
+
+        logger.debug("Got configuration info!")
+        (project_id, hash, config_dicts) = result
```
renamed everywhere to `state_hash`

manneohrstrom commented:
CR addressed. Renamed a bunch of files so earlier diff may be totally off now. Tried to address all smaller comments. The following is still needed to be addressed, but to be pragmatic, it would be great if we could do that separately if that's ok:

- Caching needs to be broken out and better encapsulated in general.
- Process handling needs to be tested against longer running QT processes. This wasn't possible before, but should be doable now. I suggest we tackle this as a separate body of work as well to keep things simple. This include process life cycle and logging/progress feedback back to the main process.

Ready for another round of CR!

sgsteph commented on python/external_config/external_config_loader.py:
```diff
@@ -0,0 +1,283 @@
+                    "state_hash": self._config_state.get_hash()
+                }
+            )
+            self._task_ids[unqiue_id] = project_id
```
`unique_id`

manneohrstrom replied:
fixed

sgsteph commented on python/external_config/external_config_loader.py:
```diff
@@ -0,0 +1,283 @@
+            "state_hash": self._config_state.get_hash()
+        }
+
+        config_data = file_cache.load_cache(config_cache_key)
```
I thought I mentioned it during my first pass, but couldn't retrieve my comment: it seems we're loading the cache for nothing if `force_reload` is `False`?

sgsteph replied:
Actually, I was writing a comment about this method being confusing, and I think I understand my confusion: `force_reload` is actually a way to force a refresh, and ignore any cached data. I wrongly associated `load` with loading the cached data, not sending a refresh request. So may be this param should be `force_refresh`, or the docstring be updated with something like:
```
       :param force_reload: If ``True``, force a refresh of the configuration data.
             If ``False`` (default), use a cached representation if available. This
             cache is refreshed at startup and whenever
             :meth:`refresh_shotgun_global_state` is called.
```
It took me a while to get it, so I think it's worth clarifying it. To me `reload` means load the cached data again.

sgsteph commented on python/external_config/scripts/external_runner.py:
```diff
@@ -0,0 +1,221 @@
+
+    :returns: Path to the current core.
+    """
+    sgtk_file = inspect.getfile(sgtk)
```
Just checking this comment was not missed?

manneohrstrom commented:
update: 
- doc build working
- no more inspect
- fixed typos
- i removed the force reload parameter, it was meant to be simple and useful but it seems more confusing than anything. Easy to add in at a later point if a need arises.
- i'm working on tickets for the rest.

manneohrstrom commented:
back for another round of CR!
